### PR TITLE
adding unlayered styles to data model

### PR DESCRIPTION
**Pull request by oluoluoxenfree**
## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

## Show me
_Provide screenshots/animated gifs/videos if necessary._

# REMEMBER: Attach this PR to the Trello card

### CODE REVIEW

File: plugins/postcss-cascade-layers/src/index.ts [hunk 1]
```diff
@@ -54,9 +54,9 @@ function postcssCascadeLayers(opts) {
 			});
 
 			// 2nd walkthrough to transform unlayered styles - need highest specificity (layerCount + 1)
-			// root.walkRules((rule) => {
-			// 	console.log("second walkthrough");
-			// });
+			root.walkRules((rule) => {
+				layerOrder['unlayered ' + rule.selector] = layerCount + 1;
+			});
```
Comment (romainmenke) [inline plugins/postcss-cascade-layers/src/index.ts]: The unlayered styles do not really need to be added to the data model individually I think.
We can just transform them directly here.

Maybe best to already add a few helper functions :

- something to generate the `:not(...)` part for a given specificity.
- something to check that a `rule` is indeed unlayered

```js
function generateNot(specificity: number) {
	let list = '';
	for (let i = 0; i < specificity; i++) {
		list += '#\\#'; // something short but still very uncommon
	}

	return `:not(${list})`;
}

function hasLayerAtRuleAncestor(node: Node): boolean {
	let parent = node.parent;
	while (parent) {
		if (parent.type !== 'atrule') {
			parent = parent.parent;
			continue;
		}

		if ((parent as AtRule).name === 'layer') {
			return true;
		}

		parent = parent.parent;
	}

	return false;
}
```

Then you can do this :

```js
			if (!layerCount) {
				// no layers, so nothing to transform.
				return;
			}

			// 2nd walkthrough to transform unlayered styles - need highest specificity (layerCount)
			root.walkRules((rule) => {
				if (hasLayerAtRuleAncestor(rule)) {
					return;
				}

				rule.selectors = rule.selectors.map((selector) => {
					// Needs `postcss-selector-parser` to insert `:not()` before any pseudo elements like `::after`
					// This is a side track and can be fixed later.
					return `${generateNot(layerCount)} ${selector}`;
				});
			});
```

Comment (oluoluoxenfree) [reply plugins/postcss-cascade-layers/src/index.ts]: What does this line mean? 

`// Needs `postcss-selector-parser` to insert `:not()` before any pseudo elements like `::after``

we need to include another module later?

@romainemenke

Comment (romainmenke) [reply plugins/postcss-cascade-layers/src/index.ts]: > we need to include another module later?

Correct.
We can only create a valid selector in each case by parsing the selectors and correctly combining all the parts :)

_update, I should not type on the move... all the typo's_

**Review by sanajaved7 — CHANGES_REQUESTED** [review]
This is great @oluoluoxenfree 🎉 Only changes I have are that it looks like there are some merge conflicts to resolve and the current tests seem to be failing. Do you think we need additional tests as well to cover the unlayered style transformation?

**Comment by sanajaved7** [comment]
[postCSS plugin](https://trello.com/c/Wj3yjQk4/16-postcss-plugin)

**Comment by oluoluoxenfree** [comment]
@sanajaved7 I think you're right about additional tests, and I'll have a look at the failures, thanks so much

**Comment by oluoluoxenfree** [comment]
@romainmenke do you know if there's any reason tape wouldn't like the not string you've set up?

It keeps adding additional slashes to the expected output when i try to put it in the expected file.

**Comment by romainmenke** [comment]
@oluoluoxenfree running `npm run test:rewrite-expects` should update the files correctly.
postcss tape is actually removing slashes I think.

**Pull request merged**